public/app/features/explore/ExploreToolbar.tsx

```diff
@@ -112,6 +118,25 @@ export class UnConnectedExploreToolbar extends PureComponent<Props, {}> {
     changeMode(exploreId, mode);
   };
 
+  returnToPanel = async () => {
+    const { originPanel, queries } = this.props;
+
+    const dashboardSrv = getDashboardSrv();
+    const dash = dashboardSrv.getCurrent();
+
+    const titleSlug = kbn.slugifyForUrl(dash.title);
+    getLocationSrv().update({
+      path: `/d/${dash.uid}/:${titleSlug}`,
+      query: {
+        fullscreen: true,
+        edit: true,
+        panelId: originPanel,
+        fromExplore: true,
+        queryRows: queries.length,
```

Comment:
it's the number of queries to take from explore history in local storage. Needed since otherwise there would be no way to know if the first, say, 3 entries in the `grafana.explore.history[dataSource]` array refer to 3 query rows at one point in history, or 1 query row at 3 points in history.
I could change the name to `queryRowCount` if that makes it more clear?

Reply:
yes that would help, as the queryRows name to me makes this variable/property represent an array of rows so the Count suffix helps make it clear it's just a number. 

So what was the reasoning behind using localstorage instead of just storing the modified query in the explore redux state? this state should not need to survive full page reloads. But maybe it aids developing and testing this feature?

Reply:
No reasoning beyond that the query state is already stored there, so I didn't think it'd be necessary to duplicate that state in redux. But if you think it's clearer just to have the modified query in redux I can change it?

Reply:
It would not be duplicate. The dashboard model is stored in redux , and the process of loading and initiation is In redux. So reading this query param and modifying the query during dashboard init is an option to keep this logic out of UI component. Worth investigating

Reply:
I guess by "duplicate" I mean having the modified query both in localstorage (via Explore history), and also having it in redux. I know that while using Explore the queries are also available from redux, but that gets reset when the Explore wrapper gets unmounted (via `resetExploreAction`), so it's no longer accessible during dashboard init. Maybe I'm misunderstanding something though?

Reply:
Ah, I missed the fact that you were using the existing local storage used by the history feature.

> I know that while using Explore the queries are also available from redux, but that gets reset when the Explore wrapper gets unmounted

I was more thinking that when the action "go back to dashboard with updated queries" action is triggered something is stored somewhere in the redux store (like the queries & fromExplore) and then the location change.  

But you could still access the query params and local storage in the dashboard init phase. That process has unit tests as well and would be easier to test I think. 
 
https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/initDashboard.ts